File: Rudder/Classes/RudderMessage.m
//
//  RudderMessage.m
//  RudderSDKCore
//
//  Created by Arnab Pal on 17/10/19.
//  Copyright © 2019 Rudderlabs. All rights reserved.
//

#import "RudderMessage.h"
#import "RudderElementCache.h"
#import "Utils.h"

@implementation RudderMessage

- (instancetype)init
{
    self = [super init];
    if (self) {
        _messageId = [[NSString alloc] initWithFormat:@"%ld-%@", [Utils getTimeStampLong], [Utils getUniqueId]];
        _channel = @"mobile";
        _context = [RudderElementCache getContext];
        _originalTimestamp = [Utils getTimestamp];
        _anonymousId = [[NSString alloc] initWithFormat:@"%@", [_context.traits objectForKey:@"anonymousId"]];
        NSObject *userIdObj = [_context.traits objectForKey:@"userId"];
        if (userIdObj != nil) {
            _userId = [[NSString alloc] initWithFormat:@"%@", userIdObj];
        }
    }
    return self;
}

- (NSDictionary<NSString*, NSObject*>*) dict {
    NSMutableDictionary* tempDict = [[NSMutableDictionary alloc] init];
    
    [tempDict setValue:_messageId forKey:@"messageId"];
    [tempDict setValue:_channel forKey:@"channel"];
    [tempDict setValue:[_context dict] forKey:@"context"];
    [tempDict setValue:_type forKey:@"type"];
    [tempDict setValue:_action forKey:@"action"];
    [tempDict setValue:_originalTimestamp forKey:@"originalTimestamp"];
    [tempDict setValue:_anonymousId forKey:@"anonymousId"];
    if (_userId != nil) {
        [tempDict setValue:_userId forKey:@"userId"];
    }
    [tempDict setValue:_properties forKey:@"properties"];
    [tempDict setValue:_event forKey:@"event"];
    [tempDict setValue:_userProperties forKey:@"userProperties"];
    [tempDict setValue:_integrations forKey:@"integrations"];
    
    return [tempDict copy];
}

- (void)updateContext:(RudderContext *)context {
    if (context != nil) {
        self.context = context;
    }
}

- (void)updateTraits:(RudderTraits *)traits {
    [_context updateTraits:traits];
}
@end

File: RudderSampleProject/RudderSdkCore/RudderMessage.swift
//
//  RudderMessage.swift
//  RudderSample
//
//  Created by Arnab Pal on 10/07/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Foundation

@objc public class RudderMessage : NSObject, Encodable {
    let messageId: String = String(NSDate().timeIntervalSince1970) + "-" + UUID().uuidString.lowercased()
    var channel: String = "mobile"
    var context : RudderContext? = nil
    var type: String = ""
    let action: String = ""
    let timestamp: String = Utils.getTimeStampStr()
    var anonymousId: String = ""
    var userId: String = ""
    var event: String = ""
    var eventProperties: AnyCodable? = nil
    var userProperties: AnyCodable? = nil
    var integrations: AnyCodable? = nil
    
    override init() {
        self.context = RudderElementCache.getCachedContext()
        self.anonymousId = self.context!.deviceInfo.id
    }
    
    func setIntegrations(integrations: AnyCodable) {
        self.integrations = integrations
    }
    
    func setIntegrations(integrations: Dictionary<String, Bool>) {
        self.integrations = AnyCodable(value: integrations)
    }
    
    func setEventProperties(eventProperties: AnyCodable) {
        self.eventProperties = eventProperties
    }
    
    func setUserProperties(userProperties: AnyCodable) {
        self.userProperties = userProperties
    }
    
    func setEventType(type: String) {
        self.type = type
    }
    
    func setEventName(eventName: String) {
        self.event = eventName
    }
    
    func setUserId(userId: String) {
        self.userId = userId
    }
    
    enum CodingKeys: String, CodingKey {
        case messageId = "messageId"
        case channel = "channel"
        case context = "context"
        case type = "type"
        case action = "action"
        case timestamp = "timestamp"
        case anonymousId = "anonymousId"
        case userId = "userId"
        case event = "event"
        case eventProperties = "properties"
        case userProperties = "userProperties"
        case integrations = "integrations"
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(messageId, forKey: .messageId)
        try container.encode(channel, forKey: .channel)
        try container.encode(context, forKey: .context)
        try container.encode(type, forKey: .type)
        try container.encode(action, forKey: .action)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(anonymousId, forKey: .anonymousId)
        try container.encode(userId, forKey: .userId)
        try container.encode(event, forKey: .event)
        try container.encode(eventProperties, forKey: .eventProperties)
        try container.encode(userProperties, forKey: .userProperties)
        try container.encode(integrations, forKey: .integrations)
    }
    
    
}
